Handle benchmarks by commit revision instead of date

philpep opened the pull request:
Commits date are not necessary monotonic along the commit log,
especially if working with rebase workflow (instead of merge), this
generate false positive or missed regression detection and unwanted
peaks in the web UI.

This patch introduce a revision number as a drop-in replacement of
commit date. This revision number is an incremental number representing
the number of ancestors commits.

In the web UI the distance between commits is the number of commits
between them. Commits can still be displayed by date using the date
selector.

Closes #390

philpep commented:
Hello,

Sorry the commit is quite huge but most of the diff is just a replacement of variables names (date -> revision).

My first attempt was to only use (sorted) commits hashes in the python code and to not use this revision number. But it was too complicated mostly due to summary graphs that need to known the "distance" between commits and the javascript code was ugly... So I just replaced the commits date by the revision number.

You can see a real demo with non-monotonic commits dates on https://hg.logilab.org/users/ppepiot/hgperf/raw-file/50a52a1968de/index.html#others.time_tags?branch=default&rev=28265&idx=2

philpep commented:
- [ ] Need to fix the web tests

philpep commented:
Ok the failed test was due to a missing sort parameter in `git log --all`. I have chosen ``--date-order`` because it seems better than ``--topo-order`` for displaying multiple branch graph (btw, both show children before parents).

pv commented:
I'd suggest using `git rev-list` instead of the UI-level command `git log`.

philpep commented:
@pv indeed, commit updated.

philpep commented on asv/www/graphdisplay.js:
```diff
@@ -1017,28 +1047,43 @@ $(document).ready(function() {
+                        point[0] = date;
+                    });
+                    graph.data.sort(function(a, b) { return a[0] - b[0]; });
+                });
```
Some hints to understand this `orig_graphs` purpose.

I was planning to use flot `transform` and `inverseTransform` functions build the `date scale` graph using `revision_to_date`. But flot assume the transform function does not reorder values (eg. is monotone), so it cannot be used here unfortunately.

So this build a new graphs and keep a reference to the old graphs for switching from one to another (but maybe you have a better idea ?).

pv commented on asv/commands/publish.py:
```diff
@@ -122,11 +109,28 @@ def run(cls, conf, env_spec=None):
                     params.setdefault(key, set())
                     params[key].add(val)
 
+            repo = get_repo(conf)
```
Probably not necessary to call `get_repo` again here?

philpep replied:
Indeed, fixed in 53388ea

pv commented:
I think the overall approach looks fine, and is the minimum perturbation. Refactoring can be thought of later, if needed.

On `orig_graphs`: looks fine. Another way would be to just call `replace_graphs()` instead of `update_graphs()` in the `#date-scale` callback function --- performance is probably not that much different, since flot is iirc the slowest part (the data is cached on js side, so no network traffic).

pv commented:
There's a small rebase needed.

pv commented on asv/graph.py:
```diff
@@ -303,21 +301,22 @@ def resample_data(val):
     # 32-bit systems can only deal with 32-bit integers, and
     # Javascript timestamps (1000*unix_timestamp) handled here
     # overflow this range
+    # FIXME: Use xrange
```
`for i in xrange(min_revision, max_revision, step_size)`  (iow just revert cc6accaee3, d88095 since it's not needed any more)

pv commented:
Also, I guess that if you get new commits in different branches, the revision numbers will change?

This will invalidate the links to the regression pages --- so maybe it would be better to use commit hashes as the `rev=` keys in those, rather than revision numbers? Otherwise, the revision numbers don't seem to be visible outside asv afaics.

philpep commented:
Hi, thanks for you feedback.

I rebased the commit and amended a revert of cc6acca (also added a test for summary graphs) and use commit hash instead of revision number in the regression url (indeed "fresh" commits can have their revision number changed if using multiple branches).

pv commented:
Thanks. Note that the web test currently fails.

philpep commented:
Yes, should be fixed now.

pv commented on asv/www/regressions.js:
```diff
@@ -181,13 +181,15 @@ $(document).ready(function() {
+            url_params.commits = [];
+            $.each(revisions, function(i, revs) {
+                var commit_b = $.asv.get_commit_hash(revs[1]);
+                if (revs[0]) {
```
`if (revs[0] === null)`? The first revision is `0` which is also false...

philpep replied:
Yes fixed (I don't known how the revision "0" could be part of a regression, but the code is more clear now).

philpep commented on asv/www/regressions.js:
```diff
@@ -364,11 +366,11 @@ $(document).ready(function() {
-        $.each(dates, function (i, date) {
-            ignore_payload = ignore_payload + ',' + date[0] + ',' + date[1];
+        $.each(revisions, function (i, revs) {
+            ignore_payload = ignore_payload + ',' + revs[0] + ',' + revs[1];
```
I think I should use commit hash here too for the html storage.

Btw this is breaking current storage based on timestamps (is this acceptable ?).

pv replied:
It's fine to change the hashes, as things become just un-ignored --- this is a per-user convenience feature (doesn't transfer between different browsers etc.).

pv commented:
Thanks, merged.

pv commented:
@philpep: did you notice the regressions here look a bit weird regarding c89b55c8: https://jenkins.philpep.org/hgperf/index.html#regressions?sort=1&dir=desc --- maybe a history flattening artefact?

philpep commented:
@pv yes I changed the hg reference repositories to benchmark against, so the results before and after c89b55c8 aren't comparables.
